File: serialize_deserialize_binary_tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return "x"
        curr = [str(root.val)]
        curr += [self.serialize(root.left)]
        curr += [self.serialize(root.right)]
        return " ".join(curr)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        def dfs(nodes):
            val = nodes.pop(0)
            if val == "x":
                return None
            currNode = TreeNode(val)
            currNode.left = dfs(nodes)
            currNode.right = dfs(nodes)
            return currNode
        if not data:
            return None
        arr = data.split(" ")
        return dfs(arr)
            
            
            
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: serialize_deserialize_binary_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
 var serialize = function(root) {
    if (!root) return 'x';
    let res = [root.val.toString()];
    res.push(serialize(root.left));
    res.push(serialize(root.right));
    return res.join(',');
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */

function buildTree(arr, idx = {val:0}) {
    if (arr.length === idx.val || arr[idx.val] === 'x') {
        return null;
    }
    let node = new TreeNode(parseInt(arr[idx.val]));
    ++idx.val;
    node.left = buildTree(arr, idx);
    ++idx.val;
    node.right = buildTree(arr, idx);
    return node;
}

var deserialize = function(data) {
    const arr = data.split(',');
    return buildTree(arr);
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */